### PR TITLE
[calyx-py] Translate `gen_exp.py` to use the new builder

### PR DESCRIPTION
**Pull request by EclecticGriffin**
Part of #1301. This includes a small bevy of minor bug-fixes to the builder including a few instances of mutable defaults being used (a particularly troublesome anti-pattern) and a lot of type annotation adjustments.

A few experiential notes:
- ran into lots of trouble with set literals for par composition as the ordering is non-deterministic and it requires the contents to be hashable which is not the case for the base ast nodes as `dataclass`es don't implement hash unless they are frozen. This means you can't use things like the `invoke` builder in a set. The ordering issue is not necessarily a problem for general use but creates issues for the runt tests so I had to hack around it in a bunch of cases. Given that, it may make more logistic sense to use a tuple for par comp instead of a set, though it doesn't really match the desired intuition. An alternative might be creating a `ParBuilder` of sorts which can be added to as desired? But in most cases that'll probably be similar to just constructing the list manually and wrapping it with the `ParComp` constructor.
- might be worth re-exposing the `ConstantPort` constructor in the builder for situations where the width cannot be inferred
- similarly, may be useful to create a `bit_low` and `bit_high` utility method for interface signals as I had to manually give widths to a bunch of these
- might be worth creating a builder method for calyx sub-components as these currently rely on constructing `CompInst` nodes manually
- some better way to do cell construction would be nice since you have to use a helper method/manually construct the node and then pass it to the cell constructor. A possible solution would be generating a lookup table for cell names? Or a standardized construction for parameter names and primitive names, similar to how the `invoke` builder uses `in_=` to set up invoke connections.
- a misc difficulty in this translation was places where the program constructed nodes which used cells that did not exist but never actually used the node. Not particularly troublesome, but did slap me in the face a handful of times. Just had to rearrange a few things to make sure the cells were only requested when the appropriate branch condition was valid.

### CODE REVIEW

**Comment by EclecticGriffin** [comment]
Just tagging @priyasrikumar in case you wanted to look things over or branch off this once the PR lands

**Comment by EclecticGriffin** [comment]
> > The ordering issue is not necessarily a problem for general use but creates issues for the runt tests so I had to hack around it in a bunch of cases.
> 
> Why not [collections.OrderedDict](https://docs.python.org/2/library/collections.html#collections.OrderedDict) ?

This is something worth considering, but it seems awkward to have to insert `none` keys since the whole point is to have a python construct whose literal we can seamlessly translate into Calyx. Possibly worth noting that the default behavior for python dicts is now ordered, so we don't need to pull out a special type if that's the way we decide to go.

**Comment by cgyurgyik** [comment]
Yeah fair enough, I was surprised Python doesn't have an ordered set (or at least I couldn't find any).

**Comment by EclecticGriffin** [comment]
Okay this now will close #1301

**Comment by EclecticGriffin** [comment]
for whatever reason github won't let me link the issues to be automatically closed. Guess I'll have to do things manually once this is merged. I'll leave it open for a bit so people have a chance to review

**Comment by rachitnigam** [comment]
@EclecticGriffin does clicking on the "Development" pane on the right and adding the issue not work?

**Comment by EclecticGriffin** [comment]
@rachitnigam it refuses to find the appropriate issue no matter what I type in. If you can make it behave go for it!

**Comment by EclecticGriffin** [comment]
Guess you have magic powers! Or github hates me specifically

**Comment by EclecticGriffin** [comment]
I'm gonna go ahead and merge this now, so as to not hold up documentation efforts

File: calyx-py/calyx/builder.py [hunk 1]
```diff
@@ -552,3 +562,14 @@ def ctx_asgn(lhs: ExprBuilder, rhs: Union[ExprBuilder, CondExprBuilder]):
...
+
+"""A one bit low signal"""
+LO = const(1, 0)
+"""A one bit high signal"""
+HI = const(1, 1)
+
+
+def par(*args: Union[GroupBuilder, ast.Control, ast.Group, ast.Invoke]) -> ast.ParComp:
+    """Build a parallel composition of control expressions."""
+    return ast.ParComp([as_control(x) for x in args])
```
Comment (sampsyo) [inline calyx-py/calyx/builder.py:575]: Awesome, and much nicer than dealing with set ordering problems.

File: calyx-py/calyx/builder.py [hunk 2]
```diff
@@ -552,3 +562,14 @@ def ctx_asgn(lhs: ExprBuilder, rhs: Union[ExprBuilder, CondExprBuilder]):
     assert TLS.groups, "assignment outside `with group`"
     group_builder: GroupBuilder = TLS.groups[-1]
     group_builder.asgn(lhs, rhs)
+
+
+"""A one bit low signal"""
+LO = const(1, 0)
+"""A one bit high signal"""
+HI = const(1, 1)
```
Comment (sampsyo) [inline calyx-py/calyx/builder.py:570]: Hell yes; the one true set of names for booleans.

**Pull request merged**